x/oracle/keeper/aggregator/filter.go

```diff
@@ -0,0 +1,85 @@
+package aggregator
+
+import (
+	"strconv"
+
+	"github.com/ExocoreNetwork/exocore/x/oracle/keeper/common"
+	"github.com/ExocoreNetwork/exocore/x/oracle/types"
+)
+
+type filter struct {
+	maxNonce int
+	maxDetID int
+	// nonce start from 1
+	validatorNonce map[string]*common.Set[int32]
+	// validator_sourceId -> roundID, NS use 0
+	validatorSource map[string]*common.Set[string]
+}
+
+func newFilter(maxNonce, maxDetID int) *filter {
+	return &filter{
+		maxNonce:        maxNonce,
+		maxDetID:        maxDetID,
+		validatorNonce:  make(map[string]*common.Set[int32]),
+		validatorSource: make(map[string]*common.Set[string]),
+	}
+}
+
+func (f *filter) newVNSet() *common.Set[int32] {
+	return common.NewSet[int32](f.maxNonce)
+}
+
+func (f *filter) newVSSet() *common.Set[string] {
```

Comment:
Consider renaming to `newValidatorNonceSet` or even `newNonceSet` and `newValidatorSourceSet` or `newSourceSet` for clarity.